README.md
[<img alt="alt_text" width="240px" src="https://cdn.buymeacoffee.com/buttons/default-orange.png" />](https://www.buymeacoffee.com/mustafahusF)

# text_mask

A package for mask textField and const text (eg. Date, phone number, IP address etc.)



<img src="https://user-images.githubusercontent.com/49743631/188289249-3512d8d5-0b5f-44cc-b9be-00c4cd7e009c.png" width="250" />


## Usage

add this line to pubspec.yaml

```yaml

   dependencies:
     text_mask: ^1.0.4

```

import package

```dart

   import 'package:text_mask/text_mask.dart';

```

## Use # for your char in text you want to mask

## Const text mask example:
```dart
  Text(
   'Phone: ${TextMask(pallet: '+90(###) ### ## ##').getMaskedText('5451312132')}',
  ),
```

## TextField Mask Example
```dart
 TextField(
  keyboardType: TextInputType.phone,
  inputFormatters: [TextMask(pallet: '+90(###) ### ## ##')],
  decoration: const InputDecoration(
   label: Text(
     'Phone',
    ),
   ),
  ),
```


CHANGELOG.md
## 1.0.4
- doc edit

## 1.0.3
- format fix

## 1.0.2
- edit readme

## 1.0.1
- edit readme

## 1.0.0
- init


lib/text_mask.dart
library text_mask;

import 'package:flutter/services.dart'
    show TextInputFormatter, TextEditingValue, TextSelection;

/// Use # for your char in text you want to mask
///
/// TextField mask example:
///  TextField(
///    keyboardType: TextInputType.phone,
///    inputFormatters: [TextMask(pallet: '(###) ### ## ##')],
///    decoration: const InputDecoration(
///     label: Text(
///      'Phone',
///    ),
///   ),
///  ),
///
///
/// Const text mask example:
///  Text(
///   'Phone: ${TextMask(pallet: '(###) ### ## ##').getMaskedText('5451316188')}',
///  ),
class TextMask extends TextInputFormatter {
  final String pallet;

  TextMask({
    required this.pallet,
  });

  @override
  TextEditingValue formatEditUpdate(
      TextEditingValue oldValue, TextEditingValue newValue) {
    String newTxt = newValue.text;
    String oldText = oldValue.text;
    newTxt = getMaskedText(newTxt, oldText: oldText);
    return TextEditingValue(
      text: newTxt,
      selection: TextSelection.collapsed(offset: newTxt.length),
    );
  }

  /// Use this for const text mask
  String getMaskedText(String newTxt, {String oldText = ''}) {
    List<Map<int, String>> list = [];
    int tempIndex = -2;
    int charAddedLength = 0;
    for (int i = 0; i < pallet.length; i++) {
      if (pallet[i] != '#') {
        if (tempIndex == i - 1) {
          list.last = {
            list.last.keys.first: list.last.values.first + pallet[i]
          };
        } else {
          list.add({i - charAddedLength: pallet[i]});
        }
        tempIndex = i;
        charAddedLength++;
      }
    }
    tempIndex = list.indexWhere((element) =>
        (oldText.endsWith(element.values.first) ||
            newTxt.endsWith(element.values.first)));
    if (tempIndex != -1 && (oldText.length > newTxt.length)) {
      newTxt = newTxt.substring(
          0, newTxt.length - list[tempIndex].values.first.length);
    } else {
      String result = '';
      for (Map<int, String> element in list) {
        newTxt = newTxt.replaceAll(element.values.first, '');
      }
      for (int i = 0; i < newTxt.length; i++) {
        final conIndex = list.indexWhere((element) => element.keys.first == i);
        if (conIndex != -1) {
          result = result + list[conIndex].values.first + newTxt[i];
        } else {
          result = result + newTxt[i];
        }
      }
      newTxt = result;
    }
    return newTxt;
  }
}


example/lib/main.dart
import 'package:flutter/material.dart';
import 'package:text_mask/text_mask.dart';

void main() {
  runApp(const MyApp());
}

class MyApp extends StatelessWidget {
  const MyApp({Key? key}) : super(key: key);

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Text Mask',
      theme: ThemeData(
        primarySwatch: Colors.blue,
      ),
      home: const MyHomePage(title: 'Text Mask'),
    );
  }
}

class MyHomePage extends StatefulWidget {
  const MyHomePage({Key? key, required this.title}) : super(key: key);

  final String title;

  @override
  State<MyHomePage> createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {
  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: Text(widget.title),
      ),
      body: Padding(
        padding: const EdgeInsets.symmetric(horizontal: 36),
        child: Column(
          crossAxisAlignment: CrossAxisAlignment.start,
          children: <Widget>[
            const SizedBox(height: 26),
            const Center(
              child: Text(
                'Mask TextField ',
                style:
                    TextStyle(color: Colors.red, fontWeight: FontWeight.bold),
                textAlign: TextAlign.center,
              ),
            ),
            TextField(
              keyboardType: TextInputType.phone,
              inputFormatters: [TextMask(pallet: '+90(###) ### ## ##')],
              decoration: const InputDecoration(
                label: Text(
                  'Phone',
                ),
              ),
            ),
            TextField(
              keyboardType: TextInputType.phone,
              inputFormatters: [TextMask(pallet: '#### #### #### ####')],
              decoration: const InputDecoration(
                label: Text(
                  'Credit Card Number',
                ),
              ),
            ),
            TextField(
              keyboardType: TextInputType.phone,
              inputFormatters: [TextMask(pallet: '##/##')],
              decoration: const InputDecoration(
                label: Text(
                  'Expiration Date',
                ),
              ),
            ),
            TextField(
              keyboardType: TextInputType.phone,
              inputFormatters: [TextMask(pallet: '###.###.###.###')],
              decoration: const InputDecoration(
                label: Text(
                  'IP Address',
                ),
              ),
            ),
            TextField(
              keyboardType: TextInputType.phone,
              inputFormatters: [TextMask(pallet: '##/##/####')],
              decoration: const InputDecoration(
                label: Text(
                  'Date',
                ),
              ),
            ),
            TextField(
              keyboardType: TextInputType.phone,
              inputFormatters: [TextMask(pallet: '# # # # # #')],
              decoration: const InputDecoration(
                label: Text(
                  'OTP',
                ),
              ),
            ),
            const SizedBox(height: 26),
            const Center(
              child: Text(
                'Mask Const Text',
                style:
                    TextStyle(color: Colors.red, fontWeight: FontWeight.bold),
                textAlign: TextAlign.center,
              ),
            ),
            const SizedBox(height: 12),
            Text(
              'Phone: ${TextMask(pallet: '+90(###) ### ## ##').getMaskedText('5451312132')}',
            ),
            const SizedBox(height: 8),
            Text(
              'Credit Card Number: ${TextMask(pallet: '#### #### #### ####').getMaskedText('1234432134567654')}',
            ),
            const SizedBox(height: 8),
            Text(
              'Ip Address: ${TextMask(pallet: '###.###.###.###').getMaskedText('123443213456')}',
            ),
          ],
        ),
      ),
    );
  }
}
